program/src/processor.rs

```diff
@@ -2479,6 +2479,167 @@ impl Processor {
+        check!(!mango_account.is_bankrupt, MangoErrorCode::Bankrupt)?;
+        check!(owner_ai.is_signer, MangoErrorCode::SignerNecessary)?;
+        check!(
+            &mango_account.owner == owner_ai.key || &mango_account.delegate == owner_ai.key,
+            MangoErrorCode::InvalidOwner
+        )?;
+
+        let open_orders_ais =
+            mango_account.checked_unpack_open_orders(&mango_group, packed_open_orders_ais)?;
+
+        let clock = Clock::get()?;
+        let now_ts = clock.unix_timestamp as u64;
+
+        let mut perp_market =
+            PerpMarket::load_mut_checked(perp_market_ai, program_id, mango_group_ai.key)?;
+        let market_index = mango_group
+            .find_perp_market_index(perp_market_ai.key)
+            .ok_or(throw_err!(MangoErrorCode::InvalidMarket))?;
+
+        let active_assets = UserActiveAssets::new(
+            &mango_group,
+            &mango_account,
+            vec![(AssetType::Perp, market_index)],
+        );
+
+        let mango_cache = MangoCache::load_checked(mango_cache_ai, program_id, &mango_group)?;
+        mango_cache.check_valid(&mango_group, &active_assets, now_ts)?;
+
+        let mut health_cache = HealthCache::new(active_assets);
+        health_cache.init_vals_with_orders_vec(
+            &mango_group,
+            &mango_cache,
+            &mango_account,
+            &open_orders_ais,
+        )?;
+        let pre_health = health_cache.get_health(&mango_group, HealthType::Init);
+
+        // update the being_liquidated flag
+        if mango_account.being_liquidated {
+            if pre_health >= ZERO_I80F48 {
+                mango_account.being_liquidated = false;
+            } else {
+                return Err(throw_err!(MangoErrorCode::BeingLiquidated));
+            }
+        }
+
+        // This means health must only go up
+        let health_up_only = pre_health < ZERO_I80F48;
+
+        let mut book = Book::load_checked(program_id, bids_ai, asks_ai, &perp_market)?;
+        let mut event_queue =
+            EventQueue::load_mut_checked(event_queue_ai, program_id, &perp_market)?;
+
+        let order = match order_type {
+            OrderType::Limit | OrderType::PostOnly | OrderType::PostOnlySlide => {
+                MangoResult::Ok(Some(Order {
+                    quantity: max_base_quantity,
+                    price: limit_price,
+                    size: 0,
+                    side,
+                }))
+            }
+            OrderType::ImmediateOrCancel | OrderType::Market => {
+                // Goes through the order book and find the price of quantity for the order
+                let depth_limit = 10;
+                if max_base_quantity > 0 {
+                    book.get_best_order_for_base_quantity(side, max_base_quantity, depth_limit);
+                }
+                if max_quote_quantity > 0 {
+                    book.get_best_order_for_quote_amount(side, max_quote_quantity, depth_limit);
```

Comment:
The return values here and above are unused?

What's the difference between `get_best_order_for_base_quantity()` and using an extreme `limit_price` with the previous order matching?

Is the idea that one can combine `max_base_quantity` and `max_quote_quantity` constraints?

Reply:
No these constraints are independants. 

On my side I was using these two function to calculate an order before it happen to see how much collateral need to be deposited on mango and estimate other things.

I then saw the interface from 01 add these parameters in they equivalent of "place_perp_order" so I gave it a go, but it's probably gibberish as it is.